File: AlgorithmCodes/src/Algorithms/Sum100.js
/**
 * Created by Rabin Shrestha on 12/25/2017.
 */

/*Algorithm for placing plus and minus signs between digits so that the result is 100

 December 7, 2013 2 Comments

 Yesterday I found the following programming problem:
 Write a program that outputs all possibilities to put + or – or nothing between the numbers 1,2,…,9 (in this order) such that the result is 100.

 This problem has 11 solutions and they are demonstrated here: Fun with numbers: place plus/minus signs between the digits.
 1 + 2 + 34 – 5 + 67 – 8 + 9 = 100
 12 + 3 – 4 + 5 + 67 + 8 + 9 = 100
 123 – 4 – 5 – 6 – 7 + 8 – 9 = 100
 123 + 4 – 5 + 67 – 89 = 100
 123 + 45 – 67 + 8 – 9 = 100
 123 – 45 – 67 + 89 = 100
 12 – 3 – 4 + 5 – 6 + 7 + 89 = 100
 12 + 3 + 4 + 5 – 6 – 7 + 89 = 100
 1 + 23 – 4 + 5 + 6 + 78 – 9 = 100
 1 + 23 – 4 + 56 + 7 + 8 + 9 = 100
 1 + 2 + 3 – 4 + 5 + 6 + 78 + 9 = 100

 HElP : https://vortexwolf.wordpress.com/2013/12/07/algorithm-plus-minus-result-100/
 */



/*======================================
*Solution 1 is :
* ======================================
*/

var digits = [1,2,3,4,5,6,7,8,9];
var searchSum = 100;

function concatPrefix(prefix, paths) {
    return paths
        .filter(function(p) { return p.length > 0; })
        .map(function(p) { return prefix.concat(p); });
}

function findPaths(sum, previousNumber, index) {
    var previousDigit = Math.abs(previousNumber%10);
    if (index >= digits.length) {
        return sum == previousNumber ? [[previousDigit]] : [];
    }

    var currentDigit = digits[index];
    var concatenatedNumber = previousNumber >= 0
        ? 10*previousNumber + currentDigit
        : 10*previousNumber - currentDigit;

    var plusPaths = findPaths(sum-previousNumber, currentDigit, index+1);
    var minusPaths = findPaths(sum-previousNumber, -currentDigit, index+1);
    var concatPaths = findPaths(sum, concatenatedNumber, index+1);

    var paths = [];
    Array.prototype.push.apply(paths, concatPrefix([previousDigit, '+'], plusPaths));
    Array.prototype.push.apply(paths, concatPrefix([previousDigit, '-'], minusPaths));
    Array.prototype.push.apply(paths, concatPrefix([previousDigit, '&'], concatPaths));
    return paths;
}

var foundPaths = findPaths(searchSum, digits[0], 1);

if (foundPaths.length == 0) {
    document.body.textContent += "no paths were found";
} else {
    for (var i = 0; i < foundPaths.length; i++) {
        var div = document.createElement("div");
        div.textContent = foundPaths[i].join("").replace(/&/g, "").replace(/(\+|\-)/g, " $1 ");
        document.body.appendChild(div);
    }
}

File: AlgorithmCodes/src/Algorithms/Sum100.java
package Algorithms;

import javax.script.ScriptEngine;
import javax.script.ScriptEngineManager;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;

/**
 * Created by Rabin Shrestha on 12/25/2017.
 * /**
 * Write a program that outputs all possibilities to put + or - or nothing
 * between the numbers 1,2,3,4,5....9(in this order) such that the result
 * is always GivenSum ( eg:100) . For example : 1+2+34-5+67-8+9=100
 *
 */
public class Sum100 {
    public static void main(String[] args)throws Exception {

         //the total inner place to put sign are : 8
        // possible sign are : 3
        // total combination : 3^8
        System.out.println(Math.pow(3,8));

        List<Character> signs= Arrays.asList('+','-','&');
        List<Integer> numbers= Arrays.asList(1,2,3,4,5,6,7,8,9);
        Integer sum=100;
        Integer numberInnerLength=numbers.size()-1;
        Double cLength=Math.pow(signs.size(),numberInnerLength);
        List<HashMap<Integer,String>> combination=new ArrayList<>();


        // pushing digit 1-9 in appropriate place leaving space for posisble sign
        for(Integer i=0;i<cLength;i++) {
            HashMap<Integer,String> vals=new HashMap<>();
            for(int j=0;j<numberInnerLength;j++) {
                vals.put(j * 2, numbers.get(j) + "");
            }
            vals.put(numberInnerLength*2,"9");
            combination.add(vals);
        }

        //Pushing the signs ( +, - or & ) in 8 possible place
        for (int k = 0; k < numberInnerLength; k++) {
            Double periodLength = cLength /( Math.pow(signs.size(), k+1));
            int signIndex = 0;
            for (int i = 0; i < cLength; i+=periodLength) {
                for (int j = 0; j < periodLength && i+j < cLength; j++) {
                    combination.get(i+j).put((k*2)+1,signs.get(signIndex)+"");
                }
                signIndex = (signIndex+1)%signs.size();
            }
        }

        // THis is the all possible combination we can have
        System.out.println(combination);
        //now check each combination for it it make up to 100 or not
        for(HashMap<Integer,String> h : combination)
        {
            Integer total=0;
            Integer prevDigit=1;
            StringBuilder exp=new StringBuilder("");
             h.values().forEach(s->exp.append(s));
             String expTemp=exp.toString().replace("&","");

            ScriptEngineManager manager = new ScriptEngineManager();
            ScriptEngine engine = manager.getEngineByName("js");
            Integer result = (Integer)engine.eval(expTemp);
            if(result==sum)
                System.out.println(expTemp);

        }

    }
    }
